standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DatabaseProduct.java

```
@@ -20,71 +20,666 @@
    * @return database product type
    */
-  public static DatabaseProduct determineDatabaseProduct(String productName) throws SQLException {
-    if (productName == null) {
-      return OTHER;
+  public static DatabaseProduct determineDatabaseProduct(String productName, Configuration c) {
+    DbType dbt;
+
+    if (theDatabaseProduct != null) {
+      Preconditions.checkState(theDatabaseProduct.dbType == getDbType(productName));
+      return theDatabaseProduct;
     }
+
+    // This method may be invoked by concurrent connections
+    synchronized (DatabaseProduct.class) {
+
+      if (productName == null) {
+        productName = UNDEFINED_NAME;
+      }
+
+      dbt = getDbType(productName);
+
+      // Check for null again in case of race condition
+      if (theDatabaseProduct == null) {
+        final Configuration conf = c!= null ? c : MetastoreConf.newMetastoreConf();
+        // Check if we are using an external database product
+        boolean isExternal = MetastoreConf.getBoolVar(conf, ConfVars.USE_CUSTOM_RDBMS);
+
+        if (isExternal) {
+          // The DatabaseProduct will be created by instantiating an external class via
+          // reflection. The external class can override any method in the current class
+          String className = MetastoreConf.getVar(conf, ConfVars.CUSTOM_RDBMS_CLASSNAME);
+          
+          if (className != null) {
+            try {
+              theDatabaseProduct = (DatabaseProduct)
+                  ReflectionUtils.newInstance(Class.forName(className), conf);
+  
+              LOG.info(String.format("Using custom RDBMS %s. Overriding DbType: %s", className, dbt));
+              dbt = DbType.CUSTOM;
+            }catch (Exception e) {
+              LOG.warn("Caught exception instantiating custom database product. Reverting to " + dbt, e);
+            }
+          }
+          else {
+            LOG.warn("Unexpected: metastore.use.custom.database.product was set, " +
```

I believe its better to throw an exception here with the appropriate message.